File: iosApp/iosApp/core/navigation/SceneNavigationHandler.swift
//
//  SceneNavigationHandler.swift
//  iosApp
//
//  Created by Bradley Phillips on 4/9/23.
//  Copyright © 2023 orgName. All rights reserved.
//

import SwiftUI
import SwiftUIRouter
import shared

class SceneNavigationHandler: SceneNavigationHandleable
{
    var iOSNavigator: Navigator?
    
    static let shared: SceneNavigationHandler = SceneNavigationHandler()
    
    var appNavigator: AppNavigationHandleable!
    
    
    init(appNavigator: AppNavigationHandleable? = nil)
    {
        self.appNavigator = appNavigator ?? NavHelper().appNavigator()
    }
    
    func setUp(navigator: Navigator? = nil)
    {
        self.iOSNavigator = navigator
        appNavigator.setSceneNavigator(newSceneNavigator: self)
    }
    
    func navigate(route: AppRoute, direction: NavigationDirection)
    {
        print("Navigating to \(route)")
        Task { await MainActor.run {iOSNavigator?.navigate(route.name) } }
    }
}

File: androidApp/src/main/java/com/megabreezy/breezybuilds_wordle/android/core/navigation/SceneNavigationHandler.kt
package com.megabreezy.breezybuilds_wordle.android.core.navigation

import androidx.navigation.NavController
import com.megabreezy.breezybuilds_wordle.core.navigation.*
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.launch

class SceneNavigationHandler(
    private val appNavigator: AppNavigationHandleable = NavHelper().appNavigator(),
    private val navController: NavController? = null,
    private val scope: CoroutineScope? = null
): SceneNavigationHandleable
{
    fun setUp()
    {
        appNavigator.setSceneNavigator(newSceneNavigator = this)
    }

    override fun navigate(route: AppRoute, direction: NavigationDirection)
    {
        try
        {
            navController?.navigate(route.name) { popUpTo(0) }
            println("Navigating to $route")
        }
        catch (e: Throwable)
        {
            scope?.launch { appNavigator.popBack(numberOfScreens = 1) }
        }
    }
}